enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,334 @@
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). This allows cluster administrators to provide their 
+  // own CA or CA bundle for validating the certificate of upstream resolvers.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key
+  // 2. The value must be a PEM encoded certificate or bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
+  
+  // serverName is the upstream server to connect to for DNS resolution. 
+  // This is required when Transport is set to "tls"
+  //
+  // +optional
+  // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}([a-zA-Z0-9-\p{L}]){2,63})$`
+  ServerName string `json:"serverName,omitempty"`
+}
+```
+
+The second API change will be to `ForwardPlugin`. `ForwardPlugin` will get 
+three new optional members:
+
+```go
+type ForwardPlugin struct {
+  // ...<snip>
+
+  // transport allows cluster administrators to opt-in to using a DNS-over-TLS 
+  // connection between cluster DNS and an upstream resolver(s). Configuring 
+  // TLS as the transport at this level without configuring a CABundle will 
+  // result in the system certificates being used to verify the serving 
+  // certificate of the upstream resolver(s).
+  //
+  // +optional
+  // +kubebuiler:default:="cleartext"
+  Transport DNSTransport `json:"transport,omitempty"`
+  
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). This allows cluster administrators to provide their 
+  // own CA or CA bundle for validating the certificate of upstream resolvers.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key
+  // 2. The value must be a PEM encoded certificate or bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
+
+  // serverName is the upstream server to connect to for DNS resolution. 
+  // This is required when Transport is set to "tls"
+  //
+  // +optional
+  // +kubebuilder:validation:Pattern=`^([a-zA-Z0-9\p{S}\p{L}]((-?[a-zA-Z0-9\p{S}\p{L}]{0,62})?)|([a-zA-Z0-9\p{S}\p{L}](([a-zA-Z0-9-\p{S}\p{L}]{0,61}[a-zA-Z0-9\p{S}\p{L}])?)(\.)){1,}([a-zA-Z0-9-\p{L}]){2,63})$`
+  ServerName string `json:"serverName,omitempty"`
+}
+```
+
+Because the list of upstreams in the `ForwardPlugin` is a `[]string`, 
+DNS-over-TLS will be attempted for all upstreams configured for a zone by 
+default and fallback to regular DNS if a TLS connection cannot be 
```

Comment:
This doesn't make sense to me.  If I want TLS, I want to be able to specify it in a way that ensures that TLS Is actually used.

Reply:
@brandisher, can we revise the semantics here?  Is this a limitation of CoreDNS?

Reply:
@Miciah this is a limitation of CoreDNS as I understand it. Here's a quote from the documentation for the forward plugin: `Also note the TLS config is “global” for the whole forwarding proxy if you need a different tls-name for different upstreams you’re out of luck.`

That coupled with this limitation `This plugin can only be used once per Server Block.` means that you could only ever define one forward block with one TLS config that applies to all upstreams per server block.

Reply:
@deads2k After chatting with @Miciah I realized that I misunderstood your comment. A connection will fail if the TLS negotiation fails but we'll be sure to cover that case with tests in the implementation.